vault/identity_store.go

```diff
@@ -621,145 +623,159 @@ func (i *IdentityStore) Invalidate(ctx context.Context, key string) {
-	// notification is for the deletion of the underlying storage entry. At
-	// this point, since all the entities belonging to this bucket are
-	// already removed, there is nothing else to be done. But, if the
-	// storage entry is non-nil, its an indication of an update. In this
-	// case, entities in the updated bucket needs to be reinserted into
-	// MemDB.
-	var entityIDs []string
 	if bucket != nil {
-		entityIDs = make([]string, 0, len(bucket.Items))
+		// The storage entry for the entity bucket exists, so we need to compare
+		// the entities in that bucket with those in MemDB and only update those
+		// that are different. The entities in the bucket storage entry are the
+		// source of truth.
+
+		// Iterate over each entity item from the bucket
 		for _, item := range bucket.Items {
-			entity, err := i.parseEntityFromBucketItem(ctx, item)
+			bucketEntity, err := i.parseEntityFromBucketItem(ctx, item)
 			if err != nil {
 				i.logger.Error("failed to parse entity from bucket entry item", "error", err)
 				return
 			}
 
-			localAliases, err := i.parseLocalAliases(entity.ID)
+			localAliases, err := i.parseLocalAliases(bucketEntity.ID)
 			if err != nil {
 				i.logger.Error("failed to load local aliases from storage", "error", err)
 				return
 			}
+
 			if localAliases != nil {
 				for _, alias := range localAliases.Aliases {
-					entity.UpsertAlias(alias)
+					bucketEntity.UpsertAlias(alias)
 				}
 			}
 
-			// Only update MemDB and don't touch the storage
-			err = i.upsertEntityInTxn(ctx, txn, entity, nil, false)
-			if err != nil {
-				i.logger.Error("failed to update entity in MemDB", "error", err)
-				return
+			var memDBEntity *identity.Entity
+			for i, entity := range memDBEntities {
+				if entity.ID == bucketEntity.ID {
+					memDBEntity = entity
+
+					// Remove this processed entity from the slice, so that
+					// all tht will be left are unprocessed entities.
+					copy(memDBEntities[i:], memDBEntities[i+1:])
+					memDBEntities = memDBEntities[:len(memDBEntities)-1]
+					break
+				}
 			}
 
-			// If we are a secondary, the entity created by the secondary
-			// via the CreateEntity RPC would have been cached. Now that the
-			// invalidation of the same has hit, there is no need of the
-			// cache. Clearing the cache. Writing to storage can't be
-			// performed by perf standbys. So only doing this in the active
-			// node of the secondary.
-			if i.localNode.ReplicationState().HasState(consts.ReplicationPerformanceSecondary) && i.localNode.HAState() == consts.Active {
-				if err := i.localAliasPacker.DeleteItem(ctx, entity.ID+tmpSuffix); err != nil {
-					i.logger.Error("failed to clear local alias entity cache", "error", err, "entity_id", entity.ID)
+			// If the entity is not in MemDB or if it is but differs from the
+			// state that's in the bucket storage entry, upsert it into MemDB.
+			if memDBEntity == nil || !reflect.DeepEqual(memDBEntity, bucketEntity) {
```

Comment:
Should we use `cmp.Equal` here instead?

I know it's typically recommended for structs due to potential equality differences in contained values (go-cmp uses `.Equals()` wherever it's implemented and defaults to the reflect.DeepEqual behavior when it isn't).

Looking a little closer, `reflect.DeepEqual` is probably fine for a few reasons:
1. I don't see any issues with equality comparison in any of the struct fields.
2. There don't appear to be any types that implement `Equals()`, so it's probably equivalent anyway?
3 The embedded timestamps are all unix epoch time.
4. We'd have to guard against unexported fields with the `cmpopts.IgnoreUnexported` option (not that big of a deal).

Reply:
It might be worthwhile to add a comment here about why this is safe (for future readers).